### PR TITLE
Fix problem with long tuples

### PR DESCRIPTION
**Pull request by gusty**
I added a failing test as first commit to demonstrate the problem, which reveals a little flaw in the current design.

Long story short, we shouldn't decode while encoding and the other way around, and we should treat types that encode to array-likes as codecs to `'Encoding []` pretty much the same way we treat dictionaries-like as `PropertyList<'Encoding>`.

This new iteration in the design affect a bit the error DU and as seen in the legacy version of `count` it proves we're on the right track, as it becomes type safer.

It also reduce the amount of methods in the interface, by assuming tuples (and unit) will encode in an array-like way.

### CODE REVIEW

File: src/Fleece.SystemJson/Fleece.SystemJson.fs [hunk 1]
```diff
@@ -97,9 +97,9 @@ type [<Struct>] Encoding = Encoding of JsonValue with
 
     static member jsonOfJsonObject (o: JsonObject) = o :> JsonValue
 
-    static member createTuple c t = function 
-        | JArray a as x -> if length a <> c then Decode.Fail.count c (Encoding x) else t a
-        | a -> Decode.Fail.arrExpected (Encoding a)
+    // static member createTuple c t = function
```
Comment (wallymathieu) [inline src/Fleece.SystemJson/Fleece.SystemJson.fs:100]: Remove?

File: src/Fleece.SystemJson/Fleece.SystemJson.fs [hunk 2]
```diff
@@ -147,29 +147,29 @@ type [<Struct>] Encoding = Encoding of JsonValue with
...
-        Encoding.createTuple 6 (fun a -> Result.map (fun a b c d e f -> (a, b, c, d, e, f)) (decoder1 a.[0]) <*> decoder2 a.[1] <*> decoder3 a.[2] <*> decoder4 a.[3] <*> decoder5 a.[4] <*> decoder6 a.[5])
-    
-    static member tuple7D (decoder1: JsonValue -> ParseResult<'a>) (decoder2: JsonValue -> ParseResult<'b>) (decoder3: JsonValue -> ParseResult<'c>) (decoder4: JsonValue -> ParseResult<'d>) (decoder5: JsonValue -> ParseResult<'e>) (decoder6: JsonValue -> ParseResult<'f>) (decoder7: JsonValue -> ParseResult<'g>) : JsonValue -> ParseResult<'a * 'b * 'c * 'd * 'e * 'f * 'g> =
-        Encoding.createTuple 7 (fun a -> Result.map (fun a b c d e f g -> (a, b, c, d, e, f, g)) (decoder1 a.[0]) <*> decoder2 a.[1] <*> decoder3 a.[2] <*> decoder4 a.[3] <*> decoder5 a.[4] <*> decoder6 a.[5] <*> decoder7 a.[6])
+    // static member unitD : JsonValue -> ParseResult<unit> =
```
Comment (wallymathieu) [inline src/Fleece.SystemJson/Fleece.SystemJson.fs:150]: Remove?

**Review by wallymathieu — APPROVED** [review]

File: src/Fleece.SystemJson/Fleece.SystemJson.fs (hunk 1 above)
Comment (gusty) [reply src/Fleece.SystemJson/Fleece.SystemJson.fs:100]: Yes, thanks for the review. This is still WIP. I should have created it as a draft but instead I stated WIP in the description.

File: src/Fleece.SystemJson/Fleece.SystemJson.fs (hunk 2 above)
Comment (gusty) [reply src/Fleece.SystemJson/Fleece.SystemJson.fs:150]: Sure

**Pull request merged**